### ISSUE TITLE
[BUG] TypeError at library import (Python 3.9.1)

### ISSUE THREAD

**Issue by ravaszf** [Intility/fastapi-azure-auth#49]
**Describe the bug**
I'm new to Python/FastAPI, so it is very well possible that it's just me doing something wrong, but my code fails to compile as soon as I add the line to import either SingleTenantAzureAuthorizationCodeBearer or MultiTenantAzureAuthorizationCodeBearer and I'm getting this error: TypeError: unhashable type: 'list'

First I thought it's related to my already written code, so I've started a clean project and followed the tutorial in the documentation, but that also fails as soon as I add the import line. So I'm completely lost here.

I'm using Python 3.9.1 and FastAPI 0.74.1

**To Reproduce**
Add code line:
`from fastapi_azure_auth import SingleTenantAzureAuthorizationCodeBearer`

**Stack trace**
```
Traceback (most recent call last):
  File "c:\program files\python39\lib\runpy.py", line 197, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "c:\program files\python39\lib\runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "C:\Users\...\AppData\Roaming\Python\Python39\Scripts\uvicorn.exe\__main__.py", line 7, in <module>
  File "C:\Users\...\AppData\Roaming\Python\Python39\site-packages\click\core.py", line 1137, in __call__
    return self.main(*args, **kwargs)
  File "C:\Users\...\AppData\Roaming\Python\Python39\site-packages\click\core.py", line 1062, in main
    rv = self.invoke(ctx)
  File "C:\Users\...\AppData\Roaming\Python\Python39\site-packages\click\core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "C:\Users\...\AppData\Roaming\Python\Python39\site-packages\click\core.py", line 763, in invoke
    return __callback(*args, **kwargs)
  File "C:\Users\...\AppData\Roaming\Python\Python39\site-packages\uvicorn\main.py", line 435, in main
    run(app, **kwargs)
  File "C:\Users\...\AppData\Roaming\Python\Python39\site-packages\uvicorn\main.py", line 461, in run
    server.run()
  File "C:\Users\...\AppData\Roaming\Python\Python39\site-packages\uvicorn\server.py", line 60, in run
    return asyncio.run(self.serve(sockets=sockets))
  File "c:\program files\python39\lib\asyncio\runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "c:\program files\python39\lib\asyncio\base_events.py", line 642, in run_until_complete
    return future.result()
  File "C:\Users\...\AppData\Roaming\Python\Python39\site-packages\uvicorn\server.py", line 67, in serve
    config.load()
  File "C:\Users\...\AppData\Roaming\Python\Python39\site-packages\uvicorn\config.py", line 458, in load
    self.loaded_app = import_from_string(self.app)
  File "C:\Users\...\AppData\Roaming\Python\Python39\site-packages\uvicorn\importer.py", line 21, in import_from_string
    module = importlib.import_module(module_str)
  File "c:\program files\python39\lib\importlib\__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1030, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1007, in _find_and_load
  File "<frozen importlib._bootstrap>", line 986, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 680, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 790, in exec_module
  File "<frozen importlib._bootstrap>", line 228, in _call_with_frames_removed
  File ".\main.py", line 7, in <module>
    from fastapi_azure_auth import SingleTenantAzureAuthorizationCodeBearer
  File "C:\Users\...\AppData\Roaming\Python\Python39\site-packages\fastapi_azure_auth\__init__.py", line 1, in <module>
    from fastapi_azure_auth.auth import (  # noqa: F401
  File "C:\Users\...\AppData\Roaming\Python\Python39\site-packages\fastapi_azure_auth\auth.py", line 20, in <module>
    class AzureAuthorizationCodeBearerBase(SecurityBase):
  File "C:\Users\...\AppData\Roaming\Python\Python39\site-packages\fastapi_azure_auth\auth.py", line 29, in AzureAuthorizationCodeBearerBase
    iss_callable: Optional[Callable[[str], Awaitable[str]]] = None,
  File "c:\program files\python39\lib\typing.py", line 262, in inner
    return func(*args, **kwds)
  File "c:\program files\python39\lib\typing.py", line 339, in __getitem__
    return self._getitem(self, parameters)
  File "c:\program files\python39\lib\typing.py", line 463, in Optional
    return Union[arg, type(None)]
  File "c:\program files\python39\lib\typing.py", line 262, in inner
    return func(*args, **kwds)
  File "c:\program files\python39\lib\typing.py", line 339, in __getitem__
    return self._getitem(self, parameters)
  File "c:\program files\python39\lib\typing.py", line 451, in Union
    parameters = _remove_dups_flatten(parameters)
  File "c:\program files\python39\lib\typing.py", line 231, in _remove_dups_flatten
    return tuple(_deduplicate(params))
  File "c:\program files\python39\lib\typing.py", line 205, in _deduplicate
    all_params = set(params)
TypeError: unhashable type: 'list'
```

**Your configuration**
Exactly the same as the tutorial here: https://intility.github.io/fastapi-azure-auth/single-tenant/fastapi_configuration

**Comment by JonasKs** [comment]
Hi!  
That’s weird. I’ll try to reproduce today or tomorrow. 

Could you let me know which version of this library you use? 

`iss_callable: Optional[Callable[[str], Awaitable[str]]] = None`  Seems to be the issue. Reading [this](https://stackoverflow.com/questions/65858528/is-collections-abc-callable-bugged-in-python-3-9-1) SO thread, an upgrade of Python 3.9 could be sufficient.

**Comment by ravaszf** [comment]
Thank you for the quick reply.

Right, I forgot to mention that I'm using the latest version of the library, 3.2.0.

I've tried to upgrade from Python 3.9.1 to 3.10.0, but that did not make any difference. Python 3.9.2 is not available for me due to company policies.

Then I changed from collections to typing and now the code compiles and I can authenticate successfully. Thanks for the suggestion.

**Comment by JonasKs** [comment]
Perfect! Would you like to submit a PR to fix it here as well? 😊

**Comment by ravaszf** [comment]
Bare in mind I'm a total noob here :) So not sure how to do that properly.

**Comment by JonasKs** [comment]
Well, I'd say that's a even better reason to make an attempt! 

I'll write the steps here, if you want to try it out (and become a contributor), you'll have the opportunity now. These steps are the same for basically all open source projects. I'll be happy to answer any questions you might have! If you don't have time or don't want to, just let me know and I'll do it. 😊 

### Fork
First of, you need to fork the project. You do this by clicking this button:

<img width="1096" alt="image" src="https://user-images.githubusercontent.com/5310116/156368329-1f30c74c-d30a-4a0f-91e6-a8413d79fd44.png">


You then go into your project, which should now exist under your own profile [here](https://github.com/ravaszf/fastapi-azure-auth). 

### Clone
Clone this project, either using `https` or `ssh`. If you haven't set up SSH keys, you chose `https`. 

<img width="920" alt="image" src="https://user-images.githubusercontent.com/5310116/156368664-5dd04109-8fad-4b0b-88cc-e0e374fc8754.png">


Copy this URL, and paste it in your terminal in the folder you'd like it to clone to:

```bash
git clone <url>
```
Example:  
<img width="578" alt="image" src="https://user-images.githubusercontent.com/5310116/156369158-68dc65cc-c2e2-4ef0-b813-954663904156.png">

**OR** with ssh:

<img width="536" alt="image" src="https://user-images.githubusercontent.com/5310116/156370867-b27bde7d-678d-4f63-9dd0-21206a313c09.png">

(Cloning projects can also be done through clients or editors, but I can't help you with that)

### Install the project dependencies
Now, we have the project installed. If you have poetry, you can just type `poetry install` and it'll install the entire environment:

<img width="578" alt="image" src="https://user-images.githubusercontent.com/5310116/156371096-82a3e7a3-6a5b-4c5c-90f8-66b79adac94d.png">


If you **don't** have poetry, you must install it first. There's the official way [here](https://python-poetry.org/docs/#installation), but if you're on a mac you can also just do a `brew install poetry`. `pip install poetry` should also work in most cases. 

When you've written `poetry install` you can write `poetry shell` to activate the environment:

<img width="569" alt="image" src="https://user-images.githubusercontent.com/5310116/156373408-133533f6-e01b-494b-8aea-b545ac1a2339.png">

### Pre-commit

This project uses pre-commit. It's not really documented anywhere, so I'll make sure to do that! How ever, all it requires is that you run `pre-commit install`. This will install the pre-commit hooks:
<img width="526" alt="image" src="https://user-images.githubusercontent.com/5310116/156373906-9c3d1ac1-ea72-48e1-9059-c51f800548ac.png">

### Changing branch and edit the code

Now, all we have to do is create a new branch (from the branch you're on) and edit the code:

```
git checkout -b fix/typing-error
```
<img width="488" alt="image" src="https://user-images.githubusercontent.com/5310116/156374374-e6325fcd-aa92-4667-8544-2379a60aadb4.png">


And then, edit our code in your editor.

Now, we can write `git status` to see what files we've made changes to:

<img width="674" alt="image" src="https://user-images.githubusercontent.com/5310116/156374600-0b2dd251-9714-4d3e-97bf-7af8d23bd176.png">

We then add this file to commit, by typing `git add <filename>`, or `git add .` to add them all. When you've added the file, you can create a commit. `git commit -m "<commit message>"` is the simplest way:

<img width="915" alt="image" src="https://user-images.githubusercontent.com/5310116/156374883-8bb68bd8-2181-4ca2-a26e-083a5977a8e3.png">

As you can see, the `pre-commit` hooks are now installing. 

### Making pre-commit hooks pass
With my changes, the pre-commit hook actually changed my file, and my commit did **not** go through:
<img width="686" alt="image" src="https://user-images.githubusercontent.com/5310116/156375014-5b10e6bb-acaa-4a70-8d15-fba9498c55e0.png">

Typing `git status` I can see that the file is changed from last time I adde it:

<img width="746" alt="image" src="https://user-images.githubusercontent.com/5310116/156375096-138718c6-850e-4752-b8f7-cc1bdd2218c4.png">

Since the file was fixed, all I need to do is to add the file and try to commit agian:

<img width="853" alt="image" src="https://user-images.githubusercontent.com/5310116/156375283-465c1bff-df1f-4e77-ba86-43126dcddea1.png">


Success!


### Pushing and creating a pull request

Now we push these changes into our branch by typing `git push origin <branch name>`:

<img width="641" alt="image" src="https://user-images.githubusercontent.com/5310116/156375408-a56a5d48-e769-4eb9-8a13-904ced5a3401.png">

Then, navigate back to this repository in the web browser and you'll see this button. Click it and create a pull request.

<img width="779" alt="image" src="https://user-images.githubusercontent.com/5310116/156375670-c2095a6f-f2d0-467b-baf6-d5c4df196c23.png">

When you create the pull request, tests will automatically run in the branch. 😊

**Comment by ravaszf** [comment]
Thanks for the detailed guide, I'd definitely would like to give it a try, but won't have time today. I'd get back to it tomorrow or the day after, as time allows. Hope that works.

**Comment by JonasKs** [comment]
No rush at all! Let me know if you have any questions or bump into any issues. 😄

**Comment by ravaszf** [comment]
Hi Jonas,

I've tried to contribute by following your guide, but ran into some issues with poetry. My company uses a proxy repo and one of the requirements of poetry is blocked, so I can't install it. Unfortunately I don1t have permissions to view the detailed report, so I don't know why is it blocked exactly, but apparently for now this is a no-go, on my company machine at least.

So I figured I'll try from my private PC, but that had nothing development related set up and got to the point where I've installed poetry with `pip install`. But then for `pre-commit install` I don't get any output. No error either, seemingly nothing happens, so not sure what's going on, but didn't want to proceed. Any ideas on that?


And another note regarding Python 3.10.0. I've made a rookie mistake and did not create venvs, and did not realize that after installing 3.10.0 my code was still running on 3.9.1. So it's actually possible that upgrading too would solve the issue. Unfortunately I can't verify as on 3.10.0 I can't install the dotenv package, due to above mentioned proxy repo blocking this as well, so I can't compile the code without rewriting it with an alternative to dotenv. Or finding the right team and getting dotenv unblocked. Neither of them would be quick.

**Comment by JonasKs** [comment]
> I've tried to contribute by following your guide, but ran into some issues with poetry. My company uses a proxy repo and one of the requirements of poetry is blocked, so I can't install it. Unfortunately I don1t have permissions to view the detailed report, so I don't know why is it blocked exactly, but apparently for now this is a no-go, on my company machine at least.

Ah, dang. 

>So I figured I'll try from my private PC, but that had nothing development related set up and got to the point where I've installed poetry with pip install. But then for pre-commit install I don't get any output. No error either, seemingly nothing happens:

Hmm, what happens if you run `pre-commit run --all-files`? 

> 

And another note regarding Python 3.10.0. I've made a rookie mistake and did not create venvs, and did not realize that after installing 3.10.0 my code was still running on 3.9.1. So it's actually possible that upgrading too would solve the issue. Unfortunately I can't verify as on 3.10.0 I can't install the dotenv package, due to above mentioned proxy repo blocking this as well, so I can't compile the code without rewriting it with an alternative to dotenv. Or finding the right team and getting dotenv unblocked. Neither of them would be quick.

Sounds pretty rough to develop at your work environment, not gonna lie. 😁

**Comment by ravaszf** [comment]
> > I've tried to contribute by following your guide, but ran into some issues with poetry. My company uses a proxy repo and one of the requirements of poetry is blocked, so I can't install it. Unfortunately I don1t have permissions to view the detailed report, so I don't know why is it blocked exactly, but apparently for now this is a no-go, on my company machine at least.
> 
> Ah, dang.
> 
> > So I figured I'll try from my private PC, but that had nothing development related set up and got to the point where I've installed poetry with pip install. But then for pre-commit install I don't get any output. No error either, seemingly nothing happens:
> 
> Hmm, what happens if you run `pre-commit run --all-files`?
Same, no output at all :/
> 
> > And another note regarding Python 3.10.0. I've made a rookie mistake and did not create venvs, and did not realize that after installing 3.10.0 my code was still running on 3.9.1. So it's actually possible that upgrading too would solve the issue. Unfortunately I can't verify as on 3.10.0 I can't install the dotenv package, due to above mentioned proxy repo blocking this as well, so I can't compile the code without rewriting it with an alternative to dotenv. Or finding the right team and getting dotenv unblocked. Neither of them would be quick.
> 
> Sounds pretty rough to develop at your work environment, not gonna lie. 😁
Yeah, especially if one's not a developer, like myself :) Also, it's a fairly new tool, Both the tool and the processes around it are still tweaked, which also does not help much.

**Comment by JonasKs** [comment]
Hmm, that’s weird. 

I’m not sure what’s wrong honestly. If you have poetry and then write `poetry install`, then `poetry shell` you should be in the environment, with pre-commit installed.

**Comment by ravaszf** [comment]
Yeah, weird. `poetry install` completed successfully, the venv shown by `poetry shell` looks a bit different than your's, but I think that's because I'm on Windows. Yet still nothing for `pre-commit install`:
![kép](https://user-images.githubusercontent.com/82642535/156574836-eaa22d36-6c4c-465d-a675-ceb8c2adc8bb.png)

Btw, it was surprisingly easy to find someone with a way around the blocked dotenv package, so I was able to test Python 3.10.0 and with that it works even if I revert to `collections.abc`.

**Comment by JonasKs** [comment]
Yeah, I honestly don’t know what’s wrong 🤔 

If you’d like to fix the bug and then submit a PR, it’s not the end of the world with linting. I can fix that in your PR 😄

**Comment by ravaszf** [comment]
Unfortunately I could not figure out what's wrong with poetry, but I've created a new PR.

**Comment by JonasKs** [comment]
Thank you! 😄  I'll fix whatever needs fixing and push a new version to PyPi ASAP.

**Comment by JonasKs** [comment]
@ravaszf , I've pushed a change to [PyPi](https://pypi.org/project/fastapi-azure-auth/).

Thank you for your bug report and contribution! Let me know if there's anything more I can help you with.
